### ISSUE TITLE
Stack trace on losing logstash connection when using customize_event

### ISSUE THREAD

**Issue by joe-urciuoli** [dwbutler/logstash-logger#97]
When the logstash service is no longer reachable for whatever reason, I get a crash in the new version. It works fine on version 16, but when I upgrade to version 19.1 I get the error pasted at the bottom of this post.

I am using Puma and my logstash configuration looks like this (it's actually a bit more complicated, but this is enough to reproduce):

```ruby
logstash_outputs = [{ type: :tcp,
                      host: ENV["LOGSTASH_HOST"],
                      port: ENV["LOGSTASH_PORT"],
                      buffer_max_items: 1,
                      error_logger: Logger.new(logstash_err_path),
                      drop_messages_on_flush_error: true}]
logstash_outputs.push({ type: :file,
                        path: config.paths['log'].first })
logger = LogStashLogger.new(
  type: :multi_delegator,
  outputs: logstash_outputs
  )
config.logger = ActiveSupport::TaggedLogging.new(logger)
      
**
LogStashLogger.configure do |config|
        config.customize_event do |event|
          event["application"] = "MyApp"
        end
end
**
```

Notes:
-buffer_max_items is set so low just for this test case (to verify our server responds well if logstash goes down)
-removing the file buffer doesn't help
-works fine on logstash version 15
-removing the bolded configuration fixes the issue, although I still get a similar error if logstash is not working on startup, but no error if it goes down thereafter

Puma caught this error:

```
Attempt to unlock a mutex which is not locked (ThreadError)
/home/gambit/.rvm/gems/ruby-2.3.1/gems/logstash-logger-0.19.1/lib/logstash-logger/buffer.rb:280:in `unlock'
/home/gambit/.rvm/gems/ruby-2.3.1/gems/logstash-logger-0.19.1/lib/logstash-logger/buffer.rb:280:in `buffer_flush'
/home/gambit/.rvm/gems/ruby-2.3.1/gems/logstash-logger-0.19.1/lib/logstash-logger/device/connectable.rb:59:in `flush'
/home/gambit/.rvm/gems/ruby-2.3.1/gems/logstash-logger-0.19.1/lib/logstash-logger/device/multi_delegator.rb:28:in `block (3 levels) in delegate'
/home/gambit/.rvm/gems/ruby-2.3.1/gems/logstash-logger-0.19.1/lib/logstash-logger/device/multi_delegator.rb:28:in `each'
/home/gambit/.rvm/gems/ruby-2.3.1/gems/logstash-logger-0.19.1/lib/logstash-logger/device/multi_delegator.rb:28:in `block (2 levels) in delegate'
/home/gambit/.rvm/gems/ruby-2.3.1/gems/logstash-logger-0.19.1/lib/logstash-logger/logger.rb:20:in `flush'
/home/gambit/.rvm/gems/ruby-2.3.1/gems/activesupport-5.0.0.1/lib/active_support/log_subscriber.rb:70:in `flush_all!'
/home/gambit/.rvm/gems/ruby-2.3.1/gems/railties-5.0.0.1/lib/rails/rack/logger.rb:43:in `call_app'
/home/gambit/.rvm/gems/ruby-2.3.1/gems/railties-5.0.0.1/lib/rails/rack/logger.rb:24:in `block in call'
/home/gambit/.rvm/gems/ruby-2.3.1/gems/activesupport-5.0.0.1/lib/active_support/tagged_logging.rb:70:in `block in tagged'
/home/gambit/.rvm/gems/ruby-2.3.1/gems/activesupport-5.0.0.1/lib/active_support/tagged_logging.rb:26:in `tagged'
/home/gambit/.rvm/gems/ruby-2.3.1/gems/activesupport-5.0.0.1/lib/active_support/tagged_logging.rb:70:in `tagged'
/home/gambit/.rvm/gems/ruby-2.3.1/gems/railties-5.0.0.1/lib/rails/rack/logger.rb:24:in `call'
/home/gambit/.rvm/gems/ruby-2.3.1/gems/request_store-1.3.1/lib/request_store/middleware.rb:9:in `call'
/home/gambit/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0.1/lib/action_dispatch/middleware/request_id.rb:24:in `call'
/home/gambit/.rvm/gems/ruby-2.3.1/gems/rack-2.0.1/lib/rack/method_override.rb:22:in `call'
/home/gambit/.rvm/gems/ruby-2.3.1/gems/rack-2.0.1/lib/rack/runtime.rb:22:in `call'
/home/gambit/.rvm/gems/ruby-2.3.1/gems/activesupport-5.0.0.1/lib/active_support/cache/strategy/local_cache_middleware.rb:28:in `call'
/home/gambit/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0.1/lib/action_dispatch/middleware/executor.rb:12:in `call'
/home/gambit/.rvm/gems/ruby-2.3.1/gems/actionpack-5.0.0.1/lib/action_dispatch/middleware/static.rb:136:in `call'
/home/gambit/.rvm/gems/ruby-2.3.1/gems/rack-2.0.1/lib/rack/sendfile.rb:111:in `call'
/home/gambit/.rvm/gems/ruby-2.3.1/gems/railties-5.0.0.1/lib/rails/engine.rb:522:in `call'
/home/gambit/.rvm/gems/ruby-2.3.1/gems/puma-3.6.0/lib/puma/configuration.rb:225:in `call'
/home/gambit/.rvm/gems/ruby-2.3.1/gems/puma-3.6.0/lib/puma/server.rb:578:in `handle_request'
/home/gambit/.rvm/gems/ruby-2.3.1/gems/puma-3.6.0/lib/puma/server.rb:415:in `process_client'
/home/gambit/.rvm/gems/ruby-2.3.1/gems/puma-3.6.0/lib/puma/server.rb:275:in `block in run'
/home/gambit/.rvm/gems/ruby-2.3.1/gems/puma-3.6.0/lib/puma/thread_pool.rb:116:in `block in spawn_thread'
```

**Comment by dwbutler** [comment]
I think I see what's going on. When a LogStashLogger logger is flushed, the buffering implementation first obtains a lock on an internal flush mutex ([See here](https://github.com/dwbutler/logstash-logger/blob/6267c82bb4ebe263add236cde4c9542e95a29963/lib/logstash-logger/buffer.rb#L215). If there is a flush error, the buffer state is reset. ([See here](https://github.com/dwbutler/logstash-logger/blob/6267c82bb4ebe263add236cde4c9542e95a29963/lib/logstash-logger/buffer.rb#L270-L271)). When the buffer state is reset, a new flush mutex is created. ([See here](https://github.com/dwbutler/logstash-logger/blob/6267c82bb4ebe263add236cde4c9542e95a29963/lib/logstash-logger/buffer.rb#L122-L123)). The flush method call then completes and attempts to unlock the flush mutex. ([See here](https://github.com/dwbutler/logstash-logger/blob/6267c82bb4ebe263add236cde4c9542e95a29963/lib/logstash-logger/buffer.rb#L279-L280)). But this is a new mutex which was never locked. Hence the error `Attempt to unlock a mutex which is not locked (ThreadError)`.

I believe the correct solution is to create a cached mutex which is always reused even when the buffer state is reset.

**Comment by dwbutler** [comment]
Try out the referenced PR and see if that fixes the issue for you.

I also happened to notice that you're setting your Rails logger to `ActiveSupport::TaggedLogging.new(logger)`. This shouldn't be necessary because LogStashLogger includes its own tagged logging implementation which is compatible with the interface Rails expects. (In fact, it was extracted from ActiveSupport.)

**Comment by dwbutler** [comment]
Fix released in 0.19.2.
